### ISSUE TITLE
.netrc taking precedent over explicitly set auth header

### ISSUE THREAD

**Issue by alukach** [encode/httpx#568]
Using `httpx==0.8`, I was trying the following:

```py
bearer_token = 'abc123'
response = await httpx.get(
    "https://url.com",
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {bearer_token}",
    },
)
print(
    f'curl "{response.request.url}" -H "Authorization: {response.request.headers["Authorization"]}\"'
)
```

However, I was surprised to see the following printed:
```
curl "https://url.com" -H "Authorization: Basic dXNlcm5hbWU6bXlTZWNyZXRQYXNzd29yZA=="
```

Going to https://gitter.im/encode/community for assistance, @florimondmanca correctly guessed that the credentials may be coming from `~/.netrc`.  This was a bit of a surprise to me. He described the current order of priority for auth as:
> Request auth parameter > URL username/password > NETRC > Request authorization header > Client authorization header

I think that, in general, explicitly set headers should never be overridden by implicit helpers.

**Comment by florimondmanca** [comment]
Just reproduced this:

- NETRC file:

```
# netrc
machine localhost:8000
login test
password test
```

- Test server (`$ uvicorn server:app`):

```python
# server.py

async def app(scope, receive, send):
    assert scope["type"] == "http"
    await receive()
    await send(
        {
            "type": "http.response.start",
            "status": 200,
            "headers": [[b"content-type", b"text/plain"]],
        }
    )
    await send({"type": "http.response.body", "body": b"Hello, world!"})
```

- Client script:

```python
# client.py
import asyncio

import httpx


async def main():
    url = "http://localhost:8000"
    resp = await httpx.get(url, headers={"Authorization": "Bearer 123"})
    print(resp.request.headers["Authorization"])


asyncio.run(main())
```

Usage:

```console
$ NETRC=path/to/netrc python client.py
Basic dGVzdDp0ZXN0  # Expected "Bearer 123"
```

**Comment by florimondmanca** [comment]
To refine the "priority order" mentioned here, which authentication source should be used should probably follow this algorithm:

1. Use the `auth` parameter passed on the request.
1. Otherwise, use the `username:password` pair encoded in the URL.
1. Otherwise, use the `Authorization` header on the request.
1. Otherwise, use the `Authorization` header on the client.
1. Otherwise, use credentials from a NETRC file.

Right now I believe we have 1 - 2 - 5 - 3 - 4.

One possibility would be to lookup NETRC at client instanciation time to build a default `Authorization` header, then let it be overridden by any `headers` passed to `Client()`.

Right now looking up NETRC is lazy (it happens on the first request, provided `trust_env=True`). I don't think making it eager is a problem since NETRC relies on environment variables and files, which I think is safe to assume they won't change until we make a request.

**Comment by florimondmanca** [comment]
Also logging here that a workaround this bug is to pass `trust_env=False` (either to the client or per-request) so that HTTPX doesn't even try to read NETRC files.

**Comment by tomchristie** [comment]
Closed via 43331cfb3d3d97c9886125b821a04891951eea81

Accidentally pushed to master, rather than via a PR. 😇
